processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java

```diff
@@ -576,7 +594,12 @@ private static ValueExtractFunction makeValueExtractFunction(
           // Add aggregations.
           final int resultRowAggregatorStart = query.getResultRowAggregatorStart();
           for (int i = 0; i < entry.getValues().length; i++) {
-            resultRow.set(resultRowAggregatorStart + i, entry.getValues()[i]);
+            if (dimsToInclude != null && groupingAggregatorsBitSet.get(i)) {
```

Comment:
Right. This is one change I am a bit anxious about. is there any existing benchmark I could use?

Reply:
I think this `if` clause is probably fine. However, if I'm reading code correctly, the new aggregators seem to do nothing and [even their result is not in use](https://github.com/apache/druid/pull/10518/files#diff-8f7ac0f08ac33a05571de6df8cc7d76932a284b94c1dec98e1b18aca7a0240bcR598) but those aggregators are still involved in hash aggregation. I'm more worrying about this because it involves serializing/deserializing the aggregator values to/from off-heap memory which is pretty expensive. Because `GroupingAggregatorFactory` is a special aggregator type which cannot be computed by regular aggregation, can we rewrite the query to not compute them in hash aggregation but add the aggregation results as what you do now?

> Right. This is one change I am a bit anxious about. is there any existing benchmark I could use?

`GroupByBenchmark` will be the easiest place for such benchmarks, but it's not probably the best place because it benchmarks the query performance of historicals. I would like to suggest to add a new one for broker performance, but this comment is not a blocker. I'm OK with adding a new benchmark in `GroupByBenchmark` for now.

Reply:
In the current Long*Aggregator implementations, there is no serialization/deserialization. Aggregations are no-op and get call just returns the value passed in the constructor. am I missing any code path here?

Reply:
I talked to @abhishekagarwal87 offline. My biggest concern is that the special handling for `GroupingAggregatorFactory` seems pretty magical because the behaviour of the factory and its aggregators is different from others. What I suggested above is making it more special but less magical, because I think it's less confusing. @abhishekagarwal87's concern is mostly around the complexity of query rewriting which requires adjusting result row signature (because the result of `grouping` function will be missing at certain points during a query after rewrite). I think we can still handle this but maybe it could be fragile because we don't have a systemic way to handle result row signature changes during a query and thus the logic to handle them will be ad-hoc. I agree with this view, so the current structure seems reasonable even though I still think, ideally, we should not involve `GroupingAggregatorFactory` in hash aggregation. Maybe we can do in the future once we have some better way to handle query writing and result row signature changes.